integration/ruler_test.go

```diff
@@ -204,6 +205,138 @@ func TestRulerAPISingleBinary(t *testing.T) {
+		exprNode.SetString(strconv.Itoa(i))
+		ruleGroupName := fmt.Sprintf("test_%d", i)
+
+		expectedGroups = append(expectedGroups,
+			NGPair{
+				Namespace: fmt.Sprintf("namespace_%d", i/numNamespaces),
+				Group:     ruleGroupName,
+			},
+		)
+
+		require.NoError(t, c.SetRuleGroup(rulefmt.RuleGroup{
+			Name:     ruleGroupName,
+			Interval: 60,
+			Rules: []rulefmt.RuleNode{{
+				Record: recordNode,
+				Expr:   exprNode,
+			}},
+		}, fmt.Sprintf("namespace_%d", i/numNamespaces)))
+	}
+
+	// Sort expectedGroups as it is currently in reverse order
+	slices.SortFunc(expectedGroups, func(a, b NGPair) int {
+		fileCompare := strings.Compare(a.Namespace, b.Namespace)
+
+		// If its 0, then the file names are the same,
+		// so compare the groups
+		if fileCompare != 0 {
+			return fileCompare
+		}
+		return strings.Compare(a.Group, b.Group)
+	})
+
+	// Wait until rulers have loaded all rules.
+	require.NoError(t, rulers.WaitSumMetricsWithOptions(e2e.Equals(numRuleGroups), []string{"cortex_prometheus_rule_group_rules"}, e2e.WaitMissingMetrics))
+
+	// Since rulers have loaded all rules, we expect that rules have been sharded
+	// between the two rulers.
+	require.NoError(t, ruler1.WaitSumMetrics(e2e.Less(float64(numRuleGroups)), "cortex_prometheus_rule_group_rules"))
+	require.NoError(t, ruler2.WaitSumMetrics(e2e.Less(float64(numRuleGroups)), "cortex_prometheus_rule_group_rules"))
+
+	// No page size limit
+	actualGroups, token, err := c.GetPrometheusRules(0, "")
+	require.NoError(t, err)
+	require.Empty(t, token)
+	require.Len(t, actualGroups, len(expectedGroups))
+	for i := 0; i < len(expectedGroups); i++ {
+		require.Equal(t, expectedGroups[i].Namespace, actualGroups[i].File)
+		require.Equal(t, expectedGroups[i].Group, actualGroups[i].Name)
+	}
+
+	// We have 9 groups, keep fetching rules with a group page size of 2. The final
+	// page should have size 1 and an empty nextToken. Also check the groups are returned
+	// in order
+	var nextToken string
+	returnedGroups := make([]NGPair, 0, len(expectedGroups))
+	for i := 0; i < 4; i++ {
+		gps, token, err := c.GetPrometheusRules(2, nextToken)
+		require.NoError(t, err)
+		require.Len(t, gps, 2)
+		require.NotEmpty(t, token)
+
+		returnedGroups = append(returnedGroups, NGPair{gps[0].File, gps[0].Name}, NGPair{gps[1].File, gps[1].Name})
+		nextToken = token
+	}
+	gps, token, err := c.GetPrometheusRules(2, nextToken)
+	require.NoError(t, err)
+	require.Len(t, gps, 1)
+	require.Empty(t, token)
+	returnedGroups = append(returnedGroups, NGPair{gps[0].File, gps[0].Name})
+
+	// Check the returned rules match the rules written
+	require.Len(t, returnedGroups, len(expectedGroups))
+	for i := 0; i < len(expectedGroups); i++ {
+		require.Equal(t, expectedGroups[i].Namespace, returnedGroups[i].Namespace)
+		require.Equal(t, expectedGroups[i].Group, returnedGroups[i].Group)
+	}
+
+	// Invalid max groups value
+	_, _, err = c.GetPrometheusRules(-1, "")
+	require.Error(t, err)
```

Comment:
We should check this returns 4xx, regressions of this type are quite common (returning 5xx instead of 4xx), so having it covered in the integration test is valuable.